### DISCUSSION TITLE
Hoppscotch Desktop for self-hosting

### DISCUSSION THREAD

**Discussion by mkohns** [hoppscotch/hoppscotch#4028] [Show and tell]
Hi all!

I understand that Hoppscotch Desktop is concentrating on stability and will care for self-hosting later.
I nevertheless really love the desktop app - and wanted to have the self-hosting figured out.
So - I took a look. In my case the SSO with Azure was in focus.

My goal:
- Having Hoppscotch Desktop running with Azure SSO and connecting to my self-hosted hoppscotch backend (local and cloud)

Actually not much was missing for this. I would like to share my thoughts about this:

The user journey
- user clicks on login

- popup opens with configured SSO providers.
 - This did not work as the platform function getAllowedAuthProviders was not implemented.
 - I just copied over the function from selfhosted-web. That worked nicely.
 - https://github.com/hoppscotch/hoppscotch/blob/eecc3db4e991ab03022167298ef78fbda405a53e/packages/hoppscotch-selfhost-web/src/platform/auth/auth.api.ts#L17

- After clicking on SSO (in my case Azure), the tauri app calls backend with /auth/microsoft?redirect_uri=desktop
https://github.com/hoppscotch/hoppscotch/blob/eecc3db4e991ab03022167298ef78fbda405a53e/packages/hoppscotch-selfhost-desktop/src/platform/auth.ts#L44 
- This at the end opens up the default browser (authorization code flow) and logs in the user. The provided redirect URI in this case is MICROSOFT_CALLBACK_URL="http://localhost:3170/v1/auth/microsoft/callback"
- After successful user login, microsoft redirect to the above URL with the auth code. This is processed here: https://github.com/hoppscotch/hoppscotch/blob/eecc3db4e991ab03022167298ef78fbda405a53e/packages/hoppscotch-backend/src/auth/auth.controller.ts#L163
 - here the trouble started, because the the implementation after retrieving all tokens and setting cookies just call this custom uri: REDIRECT_URL="hoppscotch://localhost" which opens up the hoppscotch up again. In a web scenario this would be nice as the cookies would be also transported. In a custom schema scenario as with tauri - obviously not. So I decided to place the cookie parameters just as query parameters like this:
```
  if (redirectUrl === 'desktop') {
    redirectUrl = configService.get('REDIRECT_URL');
    redirectUrl +=
      '?access_token=' +
      authTokens.access_token +
      '&refresh_token=' +
      authTokens.refresh_token;
    return res.status(HttpStatus.OK).redirect(redirectUrl);
  }
```
 - So this was nice - now the tokens are provided to the desktop app via query parameters. It was funny to see that the parsing of the query parameters was already in place. Very nice! https://github.com/hoppscotch/hoppscotch/blob/eecc3db4e991ab03022167298ef78fbda405a53e/packages/hoppscotch-selfhost-desktop/src/platform/auth.ts#L264
- The function in the end just reloads the tauri app.
- So in the local storage - the tokens have been set - but still the cookies are missing to have all running. 
- I decided to make a cookie endpoint which can set these cookies and called upfront:
```
    try {
      const accessToken = await store.get("access_token")
      const refreshToken = await store.get("refresh_token")
      if (
        accessToken &&
        accessToken.value &&
        refreshToken &&
        refreshToken.value
      ) {
        console.log("Found access token and refresh token in store")
        let url = `${import.meta.env.VITE_BACKEND_API_URL}` + "/auth/cookies"
        let params =
          "?access_token=" +
          accessToken.value +
          "&refresh_token=" +
          refreshToken.value

        const resCookie = await fetch(url + params, {
          credentials: "include",
        })
        console.log(resCookie)
      }
    } catch (error) {
      console.error("Cookie error:", error)
    }
```

The corresponding new server parts looks like this:

``` 
 @Get('cookies')
  async getCookies(@Request() req, @Res() res) {
    console.log('getCookies:', req.query);
    if (
      req.query.access_token != undefined &&
      req.query.refresh_token != undefined
    ) {
      const configService = new ConfigService();

      const currentTime = DateTime.now();
      const accessTokenValidity = currentTime
        .plus({
          milliseconds: parseInt(configService.get('ACCESS_TOKEN_VALIDITY')),
        })
        .toMillis();
      const refreshTokenValidity = currentTime
        .plus({
          milliseconds: parseInt(configService.get('REFRESH_TOKEN_VALIDITY')),
        })
        .toMillis();
      console.log('setting cookies');
      res.cookie(AuthTokenType.ACCESS_TOKEN, req.query.access_token, {
        httpOnly: true,
        path: '/',
        //secure: true,
        //sameSite: 'lax',
        //maxAge: accessTokenValidity,
      });
      res.cookie(AuthTokenType.REFRESH_TOKEN, req.query.refresh_token, {
        httpOnly: true,
        path: '/',
        //secure: true,
        //sameSite: 'lax',
        //maxAge: refreshTokenValidity,
      });
    }
    console.log('getCookie finished');
    return res.status(HttpStatus.OK).send();
  }
```

That it! 
Works like a charme even with refreshing.

I actually think that this cookie auth stuff here is not very nice.
I am thinking about replacing all this with a normal Authorization JWT Bearer in the Headers.
This would make thinks here more straightful.

Leave me your thoughts about this!

**Comment by Not-James-Bond** [comment]
Hi, So I downloaded the desktop app but If you don't mind can you tell me how & where do I need to copy these function to make oauth flow work correctly. I have my BE server in Local & Azure App Service. But I'm getting trouble Auth via Oauth2 Authorization Code Flow using Azure.

**Comment by tomcurran2** [comment]
If you're exploring self-hosting options for Hoppscotch Desktop, Hosting Mastery Hub offers a range of possibilities. By hosting Hoppscotch Desktop yourself, you gain greater control and customization over the environment. Here's how you can set it up:

1. **Select a Hosting Platform:** Choose a hosting provider that aligns with your requirements and preferences. Hosting Mastery Hub provides various hosting solutions suitable for self-hosting applications like Hoppscotch Desktop.

2. **Setup Environment:** Once you've selected your hosting platform, set up the necessary environment for hosting Hoppscotch Desktop. This may involve configuring server resources, such as CPU, memory, and storage, to meet the application's requirements.

3. **Install Dependencies:** Install any dependencies required for running Hoppscotch Desktop, such as Node.js and npm (Node Package Manager).

4. **Download and Deploy:** Obtain the Hoppscotch Desktop application files and deploy them to your hosting environment. This typically involves downloading the application source code, configuring any settings or environment variables, and starting the application.

5. **Configure Domain and SSL:** If you want to access Hoppscotch Desktop via a custom domain and secure it with SSL/TLS encryption, configure the necessary DNS settings and obtain an SSL certificate. Hosting Mastery Hub may offer tools or integrations to streamline this process.

6. **Monitor and Maintain:** Regularly monitor the performance and health of your Hoppscotch Desktop instance to ensure optimal operation. Additionally, keep the application and its dependencies up to date with the latest patches and security updates.

By self-hosting Hoppscotch Desktop through [Hosting Mastery Hub](https://hostingmasteryhub.com/), you can leverage their infrastructure and support while retaining control over your hosting environment. This allows you to tailor the setup to your specific needs and preferences, ensuring an optimal experience for managing and testing APIs.

**Reply by parasiteoflife** [reply]
Nice spam

**Comment by mkohns** [comment]
I managed on my fork to replace the cookie based authentication completely with JWT Authentication Bearers in the Header. I think this is best practise in terms of security. I left cookies in for now - to have the admin package still working. (Oh I need to say that I do not need the self-hosted-web at all. This interceptor,proxy,web thing and CORS is not easy enough for unexperienced users. I think the desktop app + backend + admin is what people need)
For now - the self hosted desktop app is working fine for windows and macos. I had some trouble with the code signing and notary - but this is solved now. I have encountered that there have been a bunch of bugs/features which have not been fixed for the desktop platform but have been addressed in the self-web package. I have migrated them and desktop is now up-to-date. 
I also changed the whole oauth2 handling in regards of the redirect URI. By now - the webkit inside tauri was used to receive the 3-legged IdP responses. This was not very nice because the user could get easily stucked. I added for both (main login, and OAuth2) the tauri-oauth2-plugin which spawns an own server for the redirection (instead of using custom scheme). This works super nice and really makes me happy.
A big thank you for the developers of hoppscotch. The architecture is well designed and getting the platform up and running was a nice journey. I will start now with the tauri 2.0 migration.
